src/transformers/trainer.py

```diff
@@ -1715,16 +1745,22 @@ def _save_checkpoint(self, model, trial, metrics=None):
                 reissue_pt_warnings(caught_warnings)
         elif is_sagemaker_mp_enabled():
             if smp.rdp_rank() == 0:
-                # Consolidate the state dict on all processed of rdp_rank 0
-                opt_state_dict = self.optimizer.state_dict()
-                # Save it and the scheduler on the main process
-                if self.args.should_save:
+            # Consolidate the state dict on all processed of rdp_rank 0
+                if self.args.smp_save_partial:
+                    opt_state_dict = self.optimizer.local_state_dict()
+                else:
+                    opt_state_dict = self.optimizer.state_dict()
```

Comment:
We wanted to standardize on gather_if_shard=False here. 

When we do that, which processes need to save the partial state dict changes.

if shard_optimizer_state, then all processes
else rdp_rank==0